### PR TITLE
Fix fillvalue str

### LINKED ISSUE #941: Setting fillvalue for datasets with variable length string dtype
This does not appear to work:
```
In [34]: f = h5py.File('varlenstr.h5')

In [35]: vlen_str = h5py.special_dtype(vlen=str)

In [36]: f.create_dataset('foo', (2,), dtype=vlen_str, fillvalue='bar')
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-36-ace807cb8186> in <module>()
----> 1 f.create_dataset('foo', (2,), dtype=vlen_str, fillvalue='bar')

~/conda/envs/xarray-netcdf4-py36/lib/python3.6/site-packages/h5py/_hl/group.py in create_dataset(self, name, shape, dtype, data, **kwds)
    106         """
    107         with phil:
--> 108             dsid = dataset.make_new_dset(self, shape, dtype, data, **kwds)
    109             dset = dataset.Dataset(dsid)
    110             if name is not None:

~/conda/envs/xarray-netcdf4-py36/lib/python3.6/site-packages/h5py/_hl/dataset.py in make_new_dset(parent, shape, dtype, data, chunks, compression, shuffle, fletcher32, maxshape, compression_opts, fillvalue, scaleoffset, track_times)
    119     if fillvalue is not None:
    120         fillvalue = numpy.array(fillvalue)
--> 121         dcpl.set_fill_value(fillvalue)
    122
    123     if track_times in (True, False):

h5py/_objects.pyx in h5py._objects.with_phil.wrapper (/Users/ilan/minonda/conda-bld/h5py_1496887972496/work/h5py/_objects.c:2846)()

h5py/_objects.pyx in h5py._objects.with_phil.wrapper (/Users/ilan/minonda/conda-bld/h5py_1496887972496/work/h5py/_objects.c:2804)()

h5py/h5p.pyx in h5py.h5p.PropDCID.set_fill_value (/Users/ilan/minonda/conda-bld/h5py_1496887972496/work/h5py/h5p.c:5073)()

h5py/h5t.pyx in h5py.h5t.py_create (/Users/ilan/minonda/conda-bld/h5py_1496887972496/work/h5py/h5t.c:17938)()

h5py/h5t.pyx in h5py.h5t.py_create (/Users/ilan/minonda/conda-bld/h5py_1496887972496/work/h5py/h5t.c:17897)()

TypeError: No conversion path for dtype: dtype('<U3')
```

Summary of the h5py configuration:
<details>
h5py    2.7.0,
HDF5    1.8.17,
Python  3.6.2 |Continuum Analytics, Inc.| (default, Jul 20 2017, 13:14:59),
[GCC 4.2.1 Compatible Apple LLVM 6.0 (clang-600.0.57)],
sys.platform    darwin,
sys.maxsize     9223372036854775807,
numpy   1.13.1
</details>

### PR DESCRIPTION
**Pull request by kmuehlbauer**
Builds on #1635, closes #941

Based on the work of @aragilar and the discussion in #941 I've added the necessary changes to get vlen string fillvalues working.

Update:

There was one test failing, but it looks like the `dtype` forwarding to `set_fill_value` make that working too. So I adapted that test. Meanwhile this is settled to prior behvior.

The changes in this PR allow string fillvalues to work for vlen strings as well as fixed size strings regardless of the encoding. One source of misunderstanding for the user could be that `bytes` and `ascii` strings have the same internal representation and will not roundtrip (since `.asstr` can't be facilitated). `fillvalue` will be returned as `bytes`. 

The following works with the current implementation.

```python
with h5py.File(f"test_xxx.h5", "w") as f:   
    dt = h5py.string_dtype()
    fill_value = 'bár'
    ds = f.create_dataset('x', (10,), dtype=dt, fillvalue=fill_value)
    assert f['x'][0] == fill_value.encode("utf-8")
    assert f['x'].asstr()[0] == fill_value
    assert f['x'].fillvalue == fill_value.encode("utf-8")

with h5py.File(f"test_xxx.h5", "w") as f:   
    dt = h5py.string_dtype("ascii")
    fill_value = b'bar'
    ds = f.create_dataset('x', (10,), dtype=dt, fillvalue=fill_value)
    assert f['x'][0] == fill_value
    assert f['x'].asstr()[0] == fill_value.decode()
    assert f['x'].fillvalue == fill_value

with h5py.File(f"test_xxx.h5", "w") as f:   
    dt = h5py.string_dtype(length=10)
    fill_value = 'bár'
    ds = f.create_dataset('x', (10,), dtype=dt, fillvalue=fill_value)
    assert f['x'][0] == fill_value.encode("utf-8")
    assert f['x'].asstr()[0] == fill_value
    assert f['x'].fillvalue == fill_value.encode("utf-8")

with h5py.File(f"test_xxx.h5", "w") as f:   
    dt = h5py.string_dtype("ascii", length=10)
    fill_value = b'bar'
    ds = f.create_dataset('x', (10,), dtype=dt, fillvalue=fill_value)
    assert f['x'][0] == fill_value
    assert f['x'].asstr()[0] == fill_value.decode("ascii")
    assert f['x'].fillvalue == fill_value
```

### CODE REVIEW

File: h5py/tests/test_dataset.py [hunk 1]
```diff
@@ -468,10 +468,12 @@ def test_compound(self):
...
 
     def test_exc(self):
-        """ Bogus fill value raises ValueError """
-        with self.assertRaises(ValueError):
-            dset = self.f.create_dataset('foo', (10,),
-                    dtype=[('a', 'i'), ('b', 'f')], fillvalue=42)
+        """ Bogus fill value converts correctly"""
+        dt = np.dtype([('a', 'i'), ('b', 'f')])
+        v = np.array([42], dtype=dt)[0]
+        dset = self.f.create_dataset('foo', (10,), dtype=[('a', 'i'), ('b', 'f')], fillvalue=42)
```
Comment (takluyver) [inline h5py/tests/test_dataset.py:474]: To my mind, this probably *should* be an error - a number is not a struct. But NumPy seems to allow a scalar value for `arr.fill(42)` and `arr[:] = 42` when arr has a compound dtype, somewhat to my surprise.

@tacaswell @aragilar any preference on this?

Comment (kmuehlbauer) [reply h5py/tests/test_dataset.py:474]: @takluyver I was surprised too. It might just work because `42` converts into `int` and `float` without problem.

**Comment by kmuehlbauer** [comment]
Sorry for the confusion. I've renamed the branch, but that was not a good idea.

**Comment by kmuehlbauer** [comment]
I've rebased on latest master.

I had to add some tweaks to properly transport fixed bytes fillvalues  through the machinery. This should be good to go for the dataset creation and reading part.

Unfortunately retrieving fillvalue via `get_fill_value` is currently broken after these changes. I'll have a look how to fix this within this PR.

**Comment by kmuehlbauer** [comment]
@takluyver This is ready for review from my side. I've updated the PR description at the top.

**Comment by kmuehlbauer** [comment]
Any chance that someone can have a look here? Review much appreciated.

File: h5py/tests/test_dataset.py (hunk 1 above)
Comment (aragilar) [reply h5py/tests/test_dataset.py:474]: I think matching numpy makes sense.

Comment (aragilar) [reply h5py/tests/test_dataset.py:474]: On the other hand, if this was failing before, I'd expect it to keep failing (unless I missed why it passes now).

**Comment by aragilar** [comment]
Apart from the change to the fillvalue test (which I'm unsure about), this seems fine to me.

**Comment by kmuehlbauer** [comment]
@aragilar I'll have a look into this.Maybe if we restrict the dtype forwarding to string types only, the former behavior can be preserved.

File: h5py/tests/test_dataset.py (hunk 1 above)
Comment (kmuehlbauer) [reply h5py/tests/test_dataset.py]: @aragilar I've limited the dtype forwarding to string types and reverted the changes to the test. This should now work the same as before.

Comment (kmuehlbauer) [reply h5py/tests/test_dataset.py]: @aragilar @takluyver Looks like my latest changes brought back the old behavior. I think we can mark this as resolved.

**Comment by kmuehlbauer** [comment]
Restricting to string dtypes fixed the test issue. Now it seems like this is ready for merge. Thanks!

**Comment by takluyver** [comment]
I've opened #2057 to keep considering the fillvalue issue, because I'm not convinced either way. But thanks for cutting out the side effects from this PR for now.

File: h5py/h5p.pyx [hunk 2]
```diff
@@ -519,13 +535,25 @@ cdef class PropDCID(PropOCID):
...
         check_numpy_write(value, -1)
+
+        # check for vlen strings
+        # create correct typeID and convert from c_str pointer to string
+        string_info = check_string_dtype(value.dtype)
+        if string_info is not None and string_info.length is None:
+            tid = py_create(value.dtype, logical=1)
+            ret = H5Pget_fill_value(self.id, tid.id, &c_ptr)
+            fill_value = c_ptr
+            value[0] = fill_value.decode(string_info.encoding, "surrogateescape")
```
Comment (takluyver) [inline h5py/h5p.pyx:551]: I would actually skip the decoding. We read all strings from datasets as bytes data by default, so returning the fill value as bytes is consistent.

Comment (kmuehlbauer) [reply h5py/h5p.pyx]: I've changed it to not decode, fixed the tests. Also updated the example code above.

**Comment by takluyver** [comment]
Thanks!

**Pull request merged**